Step2 final

ArsileLuci opened the pull request:
## Task

Design and implement a `VendingMachine` type, which behaves like a [vending machine][1]:
- `Product` should has a price and a name;
- `VendingMachine` should have a limited capacity of `Product`s;
- `VendingMachine` should be able to give a rest;
- `VendingMachine` should reject purchase if it cannot give a rest;
- `Coin`s could have only `1`, `2`, `5`, `10`, `20` and `50` nominals.

Make its usage API as convenient as you're capable to.





[Rust]: https://www.rust-lang.org

[1]: https://en.wikipedia.org/wiki/Vending_machine

ArsileLuci committed:
add base functional of VendingMachine

ArsileLuci committed:
add stock capacity

ArsileLuci committed:
marked step 2 as completed

ArsileLuci committed:
add remove empty row of products after purchase

tyranron commented on 2_idioms/src/main.rs:
```diff
@@ -1,3 +1,452 @@
+mod money {
+    use crate::trade::HasValue;
+    pub struct Amount(u32);
+    impl Into<Amount> for u32 {
```
@ArsileLuci it's more idiomatic to define `From` implementation.

Also, you may not write this manually for such trivial cases and use `derive_more` crate instead.

tyranron commented on 2_idioms/src/main.rs:
```diff
@@ -1,3 +1,452 @@
+            value += self.five_count * 5;
+            value += self.ten_count * 10;
+            value += self.twenty_count * 20;
+            value += self.fifty_count * 50;
```
@ArsileLuci why writing manually, and not user `Coin::value()`?

tyranron commented on 2_idioms/src/main.rs:
```diff
@@ -1,3 +1,452 @@
+            let two_count;
+            if lock_level <= 1 {
+                fifty_count = rest / 50;
+                rest = rest % 50;
```
@ArsileLuci again, nominals are hardcoded instead being reused.

tyranron commented on 2_idioms/src/main.rs:
```diff
@@ -1,3 +1,452 @@
+        })
+    }
+
+    pub fn insert<T: Product>(&mut self, items: Vec<Box<T>>) -> Result<(), &'static str> {
```
@ArsileLuci stringly-typed errors are not idiomatic, nor are convenient. Use enums for that and have strongly-typed errors.

tyranron commented on 2_idioms/src/main.rs:
```diff
@@ -1,3 +1,452 @@
+        let tid = TypeId::of::<T>();
+        match self.product_storage.get(&tid) {
+            None => None,
+            Some(product) => Some((product[0].value(), product.len())),
```
@ArsileLuci why just don't use `.map()` combinator instead of manual matching?

tyranron commented on 2_idioms/src/main.rs:
```diff
@@ -1,3 +1,452 @@
+        let count = amount.into().0 as usize;
+        let items = self.product_storage.get_mut(&tid).unwrap();
+
+        let drain = items.drain(0..count);
```
@ArsileLuci if I recall correctly, the `0` may be omitted in range.

tyranron commented on 2_idioms/src/main.rs:
```diff
@@ -1,3 +1,452 @@
+        for item in drain {
+            vec.push(item);
+        }
+        if items.len() == 0 {
```
@ArsileLuci use `.is_empty()` instead.

tyranron commented on 2_idioms/src/main.rs:
```diff
@@ -1,3 +1,452 @@
+    ) -> Result<(), &'static str> {
+        let mut collection = Vec::new();
+        for item in items {
+            collection.push(Box::new(item))
```
@ArsileLuci use `.collect::<Vec<_>>()` instead of manual assemblying the vector.

tyranron commented on 2_idioms/src/main.rs:
```diff
@@ -1,3 +1,452 @@
+    ) -> Result<(Vec<Box<dyn Product>>, Money), (Money, &'static str)> {
+        let items_amount = amount.into();
+        match self.stock.get_price_and_count::<T>() {
+            None => return Err((money, "there's no item of this type")),
```
@ArsileLuci use `.map_err()` and `?` operator instead of manual matching.

ArsileLuci replied:
@tyranron do you mean `.map_or()`? Because `.get_price_and_count()` rerutns `Option<(u64, usize)>` and `Option<T>` doesn't have `.map_err()` function.

tyranron replied:
@ArsileLuci yes

tyranron reviewed (changes requested):
You need to polish the implementation

ArsileLuci committed:
replace Into<Amount> on From<u32>
replace manual vec collection on map().collect()
replace hardcoded nominals with Coin::value()

ArsileLuci committed:
Merge branch 'master' into Step2_final
